sdk/python/feast/infra/materialization/snowflake_engine.py

```diff
@@ -0,0 +1,459 @@
+
+
+@dataclass
+class SnowflakeMaterializationJob(MaterializationJob):
+    def __init__(
+        self,
+        job_id: str,
+        status: MaterializationJobStatus,
+        error: Optional[BaseException] = None,
+    ) -> None:
+        super().__init__()
+        self._job_id: str = job_id
+        self._status: MaterializationJobStatus = status
+        self._error: Optional[BaseException] = error
+
+    def status(self) -> MaterializationJobStatus:
+        return self._status
+
+    def error(self) -> Optional[BaseException]:
+        return self._error
+
+    def should_be_retried(self) -> bool:
+        return False
+
+    def job_id(self) -> str:
+        return self._job_id
+
+    def url(self) -> Optional[str]:
+        return None
+
+
+class SnowflakeMaterializationEngine(BatchMaterializationEngine):
+    def update(
+        self,
+        project: str,
+        views_to_delete: Sequence[
+            Union[BatchFeatureView, StreamFeatureView, FeatureView]
+        ],
+        views_to_keep: Sequence[
+            Union[BatchFeatureView, StreamFeatureView, FeatureView]
+        ],
+        entities_to_delete: Sequence[Entity],
+        entities_to_keep: Sequence[Entity],
+    ):
+        click.echo(
+            f"Deploying materialization functions for {Style.BRIGHT + Fore.GREEN}{project}{Style.RESET_ALL}"
+        )
+        click.echo()
+
+        project_name = self.repo_config.project
+
+        path = os.path.dirname(feast.__file__)
+        copy_path = path + f"/snowflake_feast_{project_name}"
+
+        if os.path.exists(copy_path):
+            shutil.rmtree(copy_path)
+
+        copy_files = [
+            "/infra/utils/snowflake/snowpark/snowflake_udfs.py",
+            "/infra/key_encoding_utils.py",
+            "/type_map.py",
+            "/value_type.py",
+            "/protos/feast/types/Value_pb2.py",
+            "/protos/feast/types/EntityKey_pb2.py",
+        ]
+
+        package_path = copy_path + "/feast"
+        for feast_file in copy_files:
+            idx = feast_file.rfind("/")
+            if idx > -1:
+                Path(package_path + feast_file[:idx]).mkdir(parents=True, exist_ok=True)
+                feast_file = shutil.copy(
+                    path + feast_file, package_path + feast_file[:idx]
+                )
+            else:
+                feast_file = shutil.copy(path + feast_file, package_path + feast_file)
+
+        zip_path = shutil.make_archive(package_path, "zip", copy_path)
+
+        conn_config = snowflake_materialization_config(self.repo_config)
+
+        stage_path = (
+            f'"{conn_config.database}"."{conn_config.schema_}"."feast_{project_name}"'
+        )
+        with get_snowflake_conn(conn_config) as conn:
+            cur = conn.cursor()
+
+            cur.execute(f"CREATE STAGE IF NOT EXISTS {stage_path}")
+
+            cur.execute(f"PUT file://{zip_path} @{stage_path} OVERWRITE=TRUE")
+            shutil.rmtree(copy_path)
```

Comment:
probably not a huge issue if this logic doesn't happen frequently, but as is any existing materialization jobs would fail after you remove right?

Reply:
copy_path is just the path to the locally create zip file. That zip file has already been loaded into snowflake, so I am just cleaning up here. 

Maybe the one thing here that could happen is during the PUT, the overwrite happens at the same exact time someone does a materialization job and it wouldnt be able to find the code, not likely but possible

Reply:
See comment above .. by checking to see if the file already exists, we can skip this after the first run